connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/FieldPath.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.transforms.field;
+
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Struct;
+
+import java.util.Map;
+
+/**
+ * Operations to update data values and schemas based on field paths.
+ * <p>
+ * See KIP-821.
+ *
+ * @see SingleFieldPath
+ * @see MultiFieldPaths
+ */
+public interface FieldPath {
```

Comment:
True. My goal was trying to align the ops between this two implementations.
I think full alignment is achievable, and we can only offer the FieldPath interface to users: https://github.com/jeqo/kafka/pull/4
Let me know what you think. It has some nice side effects described on the internal PR.

Reply:
Hmm... I think the issue with trying to unify around a single `FieldPath` (or `FieldPaths`) interface is that we lose a lot of the guarantees that the Java type system can provide us. For example, in the PR you linked, the implementation of `MultiFieldPaths::fieldFrom` doesn't really guarantee that the user provided a single field, it just operates on a single arbitrarily-selected field, even if the user provided several fields. And I'm not sure of any case where an SMT designed to operate on a single field at a time would prefer to use an internal API designed to operate on several fields instead of a single field (e.g., `FieldsFrom` instead of `FieldFrom`).

Also, even though there's technically nothing preventing users from building against this internal API, it's still not public API and we should not encourage users to build against, e.g., the `FieldPath` interface since it's subject to change at any time, at which point plugins that depend on it would become incompatible with some versions of Kafka Connect.

Ultimately, I don't see us gaining much from this interface right now, since the ways we use the `SingleFieldPath` and `MultiFieldPaths` differ enough (at least as far as the Java type system is concerned) that it's difficult to define a single shared interface for the two of them.